core/src/main/java/com/scalar/db/api/LikeExpression.java

```diff
@@ -0,0 +1,152 @@
+      throw new IllegalArgumentException("Operator must be like or not-like");
+    }
+    if (escape == null || escape.length() > 1) {
+      throw new IllegalArgumentException(
+          "Escape character must be a string of a single character or an empty string");
+    }
+    this.escape = escape;
+    this.regexPattern =
+        convertRegexPatternFrom(
+            column.getTextValue(), escape.length() == 0 ? null : escape.charAt(0));
+  }
+
+  /**
+   * Validate and convert SQL 'like' pattern to a Java regular expression. Underscores (_) are
+   * converted to '.' and percent signs (%) are converted to '.*', other characters are quoted
+   * literally. If an escape character specified, escaping is done for '_', '%', and the escape
+   * character itself. An invalid pattern will throw {@code IllegalArgumentException}. This method
+   * is implemented referencing the following Spark SQL implementation.
+   * https://github.com/apache/spark/blob/a8eadebd686caa110c4077f4199d11e797146dc5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/StringUtils.scala
+   *
+   * @param likePattern a SQL LIKE pattern to convert
+   * @param escape an escape character.
+   * @return the equivalent Java regular expression of the given pattern
+   */
+  private String convertRegexPatternFrom(String likePattern, Character escape) {
+    if (likePattern == null) {
+      throw new IllegalArgumentException("LIKE pattern must not be null");
+    }
+
+    StringBuilder out = new StringBuilder();
+    char[] chars = likePattern.toCharArray();
+    for (int i = 0; i < chars.length; i++) {
+      char c = chars[i];
+      if (escape != null && c == escape && i + 1 < chars.length) {
+        char nextChar = chars[++i];
+        if (nextChar == '_' || nextChar == '%') {
+          out.append(Pattern.quote(Character.toString(nextChar)));
+        } else if (nextChar == escape) {
+          out.append(Pattern.quote(Character.toString(nextChar)));
+        } else {
+          throw new IllegalArgumentException("LIKE pattern must not include only escape character");
+        }
+      } else if (escape != null && c == escape) {
+        throw new IllegalArgumentException("LIKE pattern must not end with escape character");
+      } else if (c == '_') {
+        out.append(".");
+      } else if (c == '%') {
+        out.append(".*");
+      } else {
+        out.append(Pattern.quote(Character.toString(c)));
+      }
+    }
+
+    return "(?s)" + out; // (?s) enables dotall mode, causing "." to match new lines
+  }
+
+  /**
+   * Returns the escape character for LIKE operator.
+   *
+   * @return the escape character for LIKE operator
+   */
+  @Nonnull
+  public String getEscape() {
+    return escape;
+  }
+
+  /**
+   * Returns true if the LIKE pattern matches a given {@code String}.
+   *
+   * @return true if the LIKE pattern matches a given {@code String}
+   */
+  public boolean isMatchedWith(String value) {
```

Comment:
We need our own matcher in the ScalarDB layer to avoid scan-after-write properly in `Snapshot`. The matcher uses regular expressions converted from "like" patterns.